hive/src/main/java/org/apache/iceberg/hive/HiveCatalogs.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hive;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.conf.HiveConf;
+
+public final class HiveCatalogs {
+
+  private static final Cache<String, HiveCatalog> CATALOG_CACHE = Caffeine.newBuilder()
+      .weakValues()
+      .build();
+
+  private HiveCatalogs() {}
+
+  public static HiveCatalog loadCatalog(Configuration conf) {
+    // metastore URI can be null in local mode
+    String metastoreUri = conf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
+    return CATALOG_CACHE.get(metastoreUri, uri -> new HiveCatalog(conf));
```

Comment:
This might be a bit problematic as we cache the Hadoop conf. Let me think about possible implications.

Reply:
Okay, one case when this won't work is for `hadoop.` data source options.

Reply:
Building a cache `Configuration` -> `HiveCatalog` doesn't seem as an option either.

Reply:
We're okay with that for 2.4 support. This will be improved when using a real catalog for Spark 3.0. Also, the `hadoop.` options are mostly for configuring the read or write, not for controlling the metastore.

Reply:
@rdblue, you are right. However, `hadoop.` data source options can be used to set `iceberg.compress.metadata` if we want to compress metadata only in particular tables.

Ideally, `iceberg.compress.metadata` should be a table property. It is the only entry in `ConfigProperties`. The problem is that `HadoopTables` requires this config upfront to find tables. Maybe, we can circumvent this by trying files with/without `gz` suffix.